be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala

```diff
@@ -42,66 +51,86 @@ object GetMessagesByIdResponseHandler extends AskPatternConstants {
+              // only log a during last attempt
+              if (retry == 1) {
+                println("Errors: " + err.toString + "\n On:\n" + prettyPrinter(graphMessage))
+              }
+            case _ => /* DO NOTHING */
           }
         }
     }
 
-    passThroughPipeline(failedMessages, validatorFlow, remainingAttempts - 1)
+    passThroughPipeline(failedMessages.reverse, validatorFlow)
   }
 
-  // Push a set of message corresponding to one channel through the pipeline
+  /** Push a message through the given validator
+    *
+    * @param validator
+    *   The pipeline verifying the message
+    * @param message
+    *   The message to verify
+    * @param system
+    *   Implicit actor system to use the given validator
+    * @return
+    *   The result of the validator on the given message, Left for failure, Right for success
+    */
   private def messagesThroughPipeline(
       validator: Flow[GraphMessage, GraphMessage, NotUsed],
-      messageSet: Set[GraphMessage]
-  )(implicit system: ActorSystem): Map[GraphMessage, PipelineError] = {
-    var failedGraphMessages: Map[GraphMessage, PipelineError] = Map.empty
-    messageSet.foreach { message =>
-      val singleRun = Source.single(message).via(validator).runWith(Sink.foreach {
-        case Left(err: PipelineError) =>
-          failedGraphMessages += message -> err
-
-        case _ => /* Nothing to do */
-      })
-      Await.ready(singleRun, duration)
+      message: GraphMessage
+  )(implicit system: ActorSystem): GraphMessage = {
+    val output = Source.single(message).via(validator).runWith(Sink.head)
+    Await.ready(output, duration).value.get match {
+      case Success(res @ _) => res
+      case err @ _          => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, err.toString, None))
     }
-
-    failedGraphMessages
   }
 
-  private def wrapMsgInPublish(map: Map[Channel, Set[Message]]): Map[Channel, Set[GraphMessage]] = {
-    map.map { case (channel, set) =>
-      channel -> set.map(msg =>
-        Right(JsonRpcRequest(
+  /** Wrap messages in Publish RPCs along with a counter
+    *
+    * @param map
+    *   A map of channels to their set of messages
+    * @return
+    *   A list of GraphMessage - Int pairs where the GraphMessage are sorted in increasing order of their channel length and the Int is the number of retry left for each message
+    */
+  private def wrapMsgInPublish(map: Map[Channel, Set[Message]]): List[(GraphMessage, Int)] = {
+    var publishedList = map.foldLeft(List.empty[JsonRpcRequest])((acc, elem) =>
+      acc ++ elem._2.map(msg =>
+        JsonRpcRequest(
           RpcValidator.JSON_RPC_VERSION,
           MethodType.PUBLISH,
-          new Publish(channel, msg),
+          new Publish(elem._1, msg),
           Some(0)
-        ))
+        )
       )
-    }
+    )
+    publishedList = publishedList.sortWith((e1, e2) => e1.getParamsChannel.channel.length <= e2.getParamsChannel.channel.length)
```

Comment:
Is this ordering part of the decentralized specification? I find an ordering by the channel name's length quite strange... Is there some specific intent behind that?

Reply:
On a scala side, you could avoir using `var` for `publishedList` by simply returning
```scala
publishedList.sortWith(...)
.map(...)
```

Reply:
Absoluetely not in the protocol. The intent is as follows: 
In general shorter channel name are less specific and so should be treated first. 

`/root`: should be treated first as it contains the lao creation messages
`/root/lao_id`: should be treated right after since it contains the rollcalls and all.

It might not be very elegant and there are exceptions but the retries seem to be enough to deal with them